### PR TITLE
Include support to deck.gl/json to interpret string as a Javascript function

### PR DESCRIPTION
**Pull request by AdriSolid**
For #5222

- The goal is to interpret strings as a JS function in @deck.gl/json.

#### Change List
- Interpret functions, getting `@@function` key as a (function) string to convert.

#### TODO
- [x] Update documentation
- [x] Add tests

cc @alasarr

### CODE REVIEW

File: modules/json/src/helpers/parse-expression.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import {get} from '../utils/get';
+
```
Comment (Pessimistress) [inline modules/json/src/helpers/parse-expression.js:2]: Why is this renamed? Does it have anything to do with this feature?

Comment (AdriSolid) [reply modules/json/src/helpers/parse-expression.js:2]: Because we are passing `MemberExpression` parsed types and `wrapperFn` as well, so at the end we are evaluating the data string (`row[prop]`) wrapped in a function, does it make sense?

Comment (Pessimistress) [reply modules/json/src/helpers/parse-expression.js:2]: You are just adding an argument, not changing the fact that `propValue` is still a string. Undoing the renaming will significantly clean up this PR.

File: modules/json/src/helpers/parse-expression.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    if (ast.type === 'Identifier') {
+      func = row => {
+        return get(row, propValue);
+      };
+    } else if (ast.type === 'MemberExpression' && wrapperFn) {
+      func = row => {
+        return wrapperFn(expressionEval.eval(ast, row));
+      };
+    } else {
+      // NOTE: To avoid security risks, the arguments passed to the
```
Comment (Pessimistress) [inline modules/json/src/helpers/parse-expression.js:32]: This risk management is completely bypassed in your new if branch.

File: modules/json/src/json-configuration.js [hunk 3]
```diff
@@ -1,20 +1,22 @@
...
+    this.functionKey = FUNCTION_KEY;
     this.log = console; // eslint-disable-line
     this.classes = {};
     this.reactComponents = {};
     this.enumerations = {};
     this.constants = {};
+    this.functions = {};
     // TODO - this needs to be simpler, function conversion should be built in
     this.convertFunction = convertFunction;
     this.preProcessClassProps = (Class, props) => props;
```
Comment (Pessimistress) [inline modules/json/src/json-configuration.js:20]: The wrapper is unnecessary. Just set `this.convertFunction = parseExpression`

File: modules/json/src/helpers/parse-expression.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import expressionEval from 'expression-eval';
+
+const cachedExpressionMap = {
+  '-': object => object
+};
+
+// Calculates an accessor function from a JSON string or object
+// '-' : x => x
+// 'a.b.c': x => x.a.b.c
+export default function parseExpression(propValue, configuration, wrapperFn) {
```
Comment (Pessimistress) [inline modules/json/src/helpers/parse-expression.js:13]: Add JSDoc to explain what the arguments are

File: modules/json/src/json-converter.js [hunk 5]
```diff
@@ -125,6 +129,24 @@ function convertClassInstance(json, configuration) {
   return instantiateClass(type, props, configuration);
 }
 
+// Plain JS object, embed functions.
+function convertFunctionObject(json, configuration) {
+  if (configuration.functions && configuration.convertFunction) {
+    const {functionKey} = configuration;
+
+    const targetFunction = json[functionKey];
+    const dataProp = json.prop;
```
Comment (Pessimistress) [inline modules/json/src/json-converter.js:138]: `prop` is an overloaded term in deck.gl. If this only refers to a path in a data row, use `key`. If this is an expression, use `value`.

File: examples/playground/src/configuration.js [hunk 6]
```diff
@@ -33,6 +33,9 @@ export default {
     {}
   ),
 
+  // Functions that should be executed by JSON converter
+  functions: Object.assign({}),
```
Comment (Pessimistress) [inline examples/playground/src/configuration.js:37]: Just `{}`

File: modules/json/src/json-converter.js [hunk 7]
```diff
@@ -125,6 +129,24 @@ function convertClassInstance(json, configuration) {
...
+    const availableFunctions = configuration.functions;
+
+    const fn = availableFunctions[targetFunction];
+    const wrapperFn = fn(json);
+
+    return configuration.convertFunction(dataProp, configuration, wrapperFn);
```
Comment (Pessimistress) [inline modules/json/src/json-converter.js:144]: This is a very strange pattern for a user-facing configuration. Why does a user need to deal with `wrapperFn` if they are to provide their own expression evaluator? It can easily be hidden with:

```js
  return row => wrapperFn(configuration.convertFunction(dataProp, configuration));
```

File: modules/json/src/helpers/parse-expression.js (hunk 1 above)
Comment (AdriSolid) [reply modules/json/src/helpers/parse-expression.js:2]: Let's leave it like before then

File: modules/json/src/helpers/parse-expression-string.js [hunk 8]
```diff
@@ -19,23 +19,25 @@ export default function parseExpressionString(propValue, configuration) {
...
-    traverse(ast, node => {
-      if (node.type === 'CallExpression') {
-        throw new Error('Function calls not allowed in JSON expressions');
-      }
-    });
-    // TODO Something like `expressionEval.eval(ast, {row});` would be useful for unpacking arrays
-    func = row => {
-      return expressionEval.eval(ast, row);
-    };
+  if (!ast.right && !ast.left) {
```
Comment (Pessimistress) [inline modules/json/src/helpers/parse-expression-string.js:22]: Leave this file intact off you didn't intend to change its functionality. This is potentially introducing a breaking change and you need to offer the rationales behind the change.

**Comment by ajduberstein** [comment]
Nit: Can we change the title of this PR since we usually reserve RFCs for the proposal that describes a change like this?

File: examples/playground/src/app.js [hunk 9]
```diff
@@ -20,13 +20,17 @@ const INITIAL_TEMPLATE = Object.keys(JSON_TEMPLATES)[0];
...
+}
+
 function addUpdateTriggersForAccessors(json) {
   if (!json || !json.layers) return;
 
   for (const layer of json.layers) {
     const updateTriggers = {};
     for (const [key, value] of Object.entries(layer)) {
-      if (key.startsWith('get') && typeof value === 'string') {
+      if ((key.startsWith('get') && typeof value === 'string') || isFunctionObject(value)) {
```
Comment (ibgreen) [inline examples/playground/src/app.js:33]: I think `@@function` syntax could be supported as generic "hydration" syntax on the top level - not just for accessors - as an equal citizen with `@@type`. 

We could also support `@@class` as an alternative to `@@type` - so you could either hydrate class instances or functions.

(A third option could also just be to have `@@type` support both - look into the JSON configuration and if there is a function registered under that name use that, if there is a class instantiate that. But having separate keyword may make things a little easier for the users).

**Review by ibgreen — APPROVED** [review]

File: examples/playground/src/app.js (hunk 9 above)
Comment (alasarr) [reply examples/playground/src/app.js:33]: +1, we should do extend this outside of accessors and It was the idea of the original [RFC](https://github.com/visgl/deck.gl/issues/5222)

Comment (alasarr) [reply examples/playground/src/app.js:33]: I'm more fan of creating a new @@function and keep the old as `@@type`. Open to have `@class` as an alternative (it's now a better name), but not a priority for me right now.

**Review by alasarr — CHANGES_REQUESTED** [review]
Add support for non-accessores

File: modules/json/src/json-converter.js [hunk 10]
```diff
@@ -125,6 +129,25 @@ function convertClassInstance(json, configuration) {
...
+function convertFunctionObject(json, configuration) {
+  if (configuration.functions && configuration.convertFunction) {
+    const {functionKey} = configuration;
+
+    const targetFunction = json[functionKey];
+    const dataProp = json.key;
+    const availableFunctions = configuration.functions;
+
+    const fn = availableFunctions[targetFunction];
+    const wrapperFn = fn(json);
```
Comment (alasarr) [inline modules/json/src/json-converter.js:142]: Need to remove the @@function key

File: modules/json/src/json-converter.js [hunk 11]
```diff
@@ -125,6 +129,22 @@ function convertClassInstance(json, configuration) {
...
 }
 
+// Plain JS object, embed functions.
+function convertFunctionObject(json, configuration) {
+  const {functionKey} = configuration;
+
+  let props = {...json};
+  delete props[functionKey];
+
+  props = convertPlainObject(props, configuration);
```
Comment (alasarr) [inline modules/json/src/json-converter.js:139]: Simpler and more generic 👌

**Review by alasarr — CHANGES_REQUESTED** [review]
Needs to update the doc of JSON/module

**Comment by Pessimistress** [comment]
Can you add at least one test case before landing?

File: modules/json/src/json-converter.js [hunk 12]
```diff
@@ -125,6 +129,22 @@ function convertClassInstance(json, configuration) {
...
+
+  const targetFunction = json[functionKey];
+  const availableFunctions = configuration.functions;
+  const matchedFn = availableFunctions[targetFunction];
```
Comment (Pessimistress) [inline modules/json/src/json-converter.js:143]: There needs to be a safety check that `matchedFn` actually exists

**Review by ibgreen — COMMENTED** [review]
I'd like to see an update of the JSON module docs. Mention this syntax with an example.

And at lease one test case.

**Comment by AdriSolid** [comment]
Tests added! :grinning:

File: docs/api-reference/json/conversion-reference.md [hunk 13]
```diff
@@ -56,6 +57,61 @@ like that in the configuration object above. These classes can be layers, views,
...
+`JSONConverter`–
+
+```js
+function calculateRadius({base, exponent}) {
+  return Math.pow(base, exponenet);
+}
+
+const configuration = {
+  ...,
+  functions: Object.assign({calculateRadius})
```
Comment (alasarr) [inline docs/api-reference/json/conversion-reference.md:72]: Need  to update https://github.com/visgl/deck.gl/blob/master/docs/api-reference/json/json-configuration.md

File: docs/api-reference/json/conversion-reference.md [hunk 14]
```diff
@@ -56,6 +57,61 @@ like that in the configuration object above. These classes can be layers, views,
...
 
 
+## Functions and using `@@function`
+
+Any JavaScript function can be passed. For example, when this configuration of functions is passed to a
+`JSONConverter`–
+
+```js
+function calculateRadius({base, exponent}) {
+  return Math.pow(base, exponenet);
```
Comment (alasarr) [inline docs/api-reference/json/conversion-reference.md:67]:
```suggestion
  return Math.pow(base, exponent);
```

**Review by alasarr — APPROVED** [review]

File: docs/api-reference/json/json-converter.md [hunk 15]
```diff
@@ -83,6 +83,55 @@ will replace the layers descriptor with
...
+A map of functions that should be made available to the JSON function resolver. For example, when this configuration is passed to `JSONConverter`:
+
+```js
+function calculateRadius({base, exponent}) {
+  return Math.pow(base, exponenet);
+}
+
+const configuration = {
+  ...,
+  functions: Object.assign({calculateRadius})
```
Comment (Pessimistress) [inline docs/api-reference/json/json-converter.md:97]: What is the purpose of `Object.assign`?

Comment (AdriSolid) [reply docs/api-reference/json/json-converter.md:97]: We can set this like `functions: {calculateRadius}`
